ocaml/libs/xapi-rrd/lib/rrd.ml

```diff
@@ -744,7 +744,19 @@ let from_xml input =
           let name = get_el "name" i in
           let type_ = get_el "type" i in
           let min_hb = get_el "minimal_heartbeat" i in
-          let min = get_el "min" i in
+          (* CA-408126 - work around negative data in historical RRDs
+             where ds_min could have been incorrectly set to neg_infinity.
+             Currently the only metric that could be negative is
+             DCMI-power-reading, so account for it.
+             Setting ds_min to 0. means Fring.make below will turn negative
+             historical values to NaNs.*)
+          let min = float_of_string (get_el "min" i) in
+          let min =
+            if min < 0. && name <> "DCMI-power-reading" then
```

Comment:
I'm not supper happy about this specific hardcoding. Can this be done by type of aggregation (only gauges are allowed to be negative, for example)

Reply:
Not really, the DCMI example is a Gauge, but other Gauges were affected. I could add a quality-gate check that checks if no other datasources supply non-zero values to the ~min parameter.

Reply:
I don't think that will be necessary. Maybe we need some regression test to ensure out-of-range values don't enter the archives.

Reply:
min is 0 for DCMI-power-reading though, and max is 65534. I don't see why it needs special casing.

Reply:
It's min_float, which is apparently the smallest positive non-zero float value. Will drop this on Monday.

Reply:
hmm. 'xe host-data-source-list' reports the wrong thing then, it reports it as 0 :(

Reply:
But anyway negative power usage doesn't make sense, so we should fix rrdp_dcmi.ml to use 0 as well in the plugin itself for the minimum.